ouroboros-network-framework/test/Test/Simulation/Network/Snocket.hs

```diff
@@ -338,30 +354,299 @@ toBearerInfo abi =
+            (\fd -> do
+              bind snocket fd localAddress
+              listen snocket fd
+              mbFd <- acceptF (accept snocket fd)
+              mbFd' <- assertState mbFd
+              traverse_ closeF mbFd'
+              return (void mbFd')
+            )
+
+acceptOne :: MonadMask m => m (Accept m fd addr) -> m (Either SomeException fd)
+acceptOne accept = mask_ $ do
+  accept0 <- accept
+  (accepted, _) <- runAccept accept0
+  case accepted of
+    Accepted fd' _ -> do
+      return (Right fd')
+    AcceptFailure err ->
+      return (Left err)
+
+runClient
+  :: ( MonadThread m
+     , MonadCatch m
+     )
+  => addr -- ^ Local Address
+  -> addr -- ^ Remote Address
+  -> Snocket m fd addr
+  -> (fd -> m ()) -- ^ Resource cleanup function (socket close)
+  -> m (Either SomeException ())
+runClient localAddress remoteAddress snocket closeF = do
+    labelThisThread "client"
+    bracket (openToConnect snocket localAddress)
+            closeF
+            $ \fd -> do
+              bind snocket fd localAddress
+              Right <$> connect snocket fd remoteAddress
+              `catch` (\(e :: SomeException) -> return (Left e))
+
+listenAndConnect
+  :: ( MonadThread m
+     , MonadCatch m
+     , Ord addr
+     )
+  => TestAddress addr -- ^ Local Address
+  -> TestAddress addr -- ^ Remote Address
+  -> Snocket m fd (TestAddress addr)
+  -> m (ObservableNetworkState (TestAddress addr))
+  -> m (Either SomeException ())
+listenAndConnect localAddress remoteAddress snocket getState = do
+    labelThisThread "connectingServer"
+    bracket (open snocket TestFamily)
+            (close snocket)
+            $ \fd -> do
+              bind snocket fd localAddress
+              listen snocket fd
+              bracket (openToConnect snocket localAddress)
+                      (close snocket)
+                      $ \fd' -> do
+                        bind snocket fd' localAddress
+                        res <- (Right <$> connect snocket fd' remoteAddress)
+                              `catch` (\(e :: SomeException) -> return (Left e))
+                        assertNetworkState localAddress remoteAddress
+                                            getState res
+
+-- | Asserts that the local address and remote address pair exists in the
+-- NetworkState.
+assertNetworkState :: (Monad m, Ord addr)
+                    => addr -- ^ Local Address
+                    -> addr -- ^ Remote Address
+                    -> m (ObservableNetworkState addr)
+                    -> Either SomeException b
+                    -> m (Either SomeException b)
+assertNetworkState localAddress remoteAddress getState res = do
+  us <- onsConnections <$> getState
+  -- Important to use serverAddr as first argument
+  let normalisedId = normaliseId localAddress remoteAddress
+  case Map.lookup normalisedId us of
+    Just _ -> return res
+    Nothing -> case res of
+      Right _ ->
+        return (Left (toException UnexpectedOutcome))
```

Comment:
we should return something more descriptive than `UnexpectedOutcome`.

Reply:
fixed!